% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/primer_design_functions.R
\name{design_primers}
\alias{design_primers}
\title{Primer design}
\usage{
design_primers(
  sequences,
  length_max = 30,
  length_min = 16,
  ROI = NA,
  min_dist = 2,
  max_dist = 50,
  gc_thres = 30,
  amplicon_length_opt = 90,
  mismatch_tolerance = 0,
  Tm_in = NA,
  Tm_delta_max = 1.5,
  Tm_delta_probe = 5,
  n_sets = 200,
  Na = 300,
  primer_conc = 0.5,
  mismatch_threshold = 0.05,
  max_repetitions = 3,
  max_complementarity = 4,
  check_spec = NA,
  position_probe = NA,
  seqs_aligned = FALSE
)
}
\arguments{
\item{sequences}{Input sequences (DNAStringSet object) that primers should be designed for}

\item{length_max}{Maximum length of primers to be considered. Defaults to 30 nt.}

\item{length_min}{Minimum length of primers to be considered. Defaults to 16 nt.}

\item{ROI}{Numeric vector of format c(start, end) that specifies region of interest within target sequence. Defaults to NA (whole sequence is examined)}

\item{min_dist}{Minimum distance between primers and probes. Defaults to 2 nt.}

\item{max_dist}{Maximum distance between primers and probes. Defaults to 50 nt. WARNING: strongly increases computation time}

\item{gc_thres}{Minimum GC content (in \%) that any sequence must have to be considered.}

\item{amplicon_length_opt}{Optimal amplicon length. Defaults to 90 nt.}

\item{mismatch_tolerance}{How many ambiguous bases will be tolerated in any primer/probe sequence. Increase only for highly heterogeneous data sets.}

\item{Tm_in}{Target melting temperature for primers. Defaults to NA, then any primers with Tm >50 degree C and Tm<65 degree C will be considered.}

\item{Tm_delta_max}{Maximum Tm difference between forward and reverse primer that will be tolerated.}

\item{Tm_delta_probe}{Minimum difference between primer Tm and probe Tm in any given primer-probe set. Defaults to 5 degree C.}

\item{n_sets}{How many primer-probe sets to return. Defaults to 200.}

\item{Na}{Sodium concentration in mM to be assumed for Tm calculation. Defaults to 300 mM. If set to a non-numeric value, Tm will be calculated based on GC content.}

\item{primer_conc}{Primer concentration in microM to be assumed for Tm calculation. Defaults to 0.5 microM. If set to a non-numeric value, Tm will be calculated based on GC content.}

\item{mismatch_threshold}{Threshold of which fraction of the input sequences can disagree at one specific base for it still to be considered a clear consensus. Defaults to 0.05.}

\item{max_repetitions}{Maximum number of repetitions of the same (di-) nucleotide in any primer/probe sequence. Defaults to 3.}

\item{max_complementarity}{Maxmium alignment score with self or other primers of the same set. Defaults to 4.}

\item{position_probe}{If given (as nt from start of the complete sequence), only probes covering this position in the sequence will be considered}

\item{seqs_aligned}{Whether input sequences are aligned. Setting to TRUE if sequences are already aligned saves computation time.}
}
\value{
list of two elements:

all_seqs: data.frame with unique ID, start position within sequence (pos), sequence information (seq), melting Temperature (Tm), GC content (GC), number of ambivalent bases (mismatches), self complementarity score (self_complementary) and self 3prime complementarity score (self_3prime_complementarity)
}
\description{
This function allows you to design qPCR/ddPCR primer and probe sets that fit to a large set of related nucleic acid sequences.
Tm is calculated using the nearest neighbor method.
}
\details{
This function searches for primer-probe sets within the input sequences that fit without any ambiguous base pairings to as many of the sequences as possible. It first aligns the input sequences and determines the consensus sequence, then it produces a list of all possible primers that could be found in the target sequence. This list is then filtered to eliminate any oligonucleotides that would not be suitable as primers(too many ambiguous bases, wrong melting temperature, too low GC content ...). Then, for each primer combinations with a probe and another primer are determined and examined in accordance with the input parameters (difference in melting temperature, position, ...). The function then gives a list of potential primer-probe sets with a score that gives an approximate measurement of the performance of the primer-probe set in qPCR/ddPCR.
}
\keyword{primer}
